include/graph_compiler.h

```diff
@@ -0,0 +1,93 @@
+#ifndef GRAPH_COMPILER_H
+
+#include <cstddef>
+#include <cinttypes>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct graph_compiler;
+struct graph_compiler_executable;
+
+struct graph_compiler_context {
+    uint32_t num_threads;
+    // ...
+};
+
+struct graph_compiler_tensor {
+    size_t id;
+    uint8_t ndims;
+    size_t dims[12];
```

Comment:
```suggestion
    size_t *dims;
```
As far as I remember the main concern against using a pointer here was that a user would need to allocate memory for each tensor only to describe dims. However, in the case of oneDNN, they already defined dims and memory for them via oneDNN's tensors, so why wouldn't we simply reuse the same memory?

The main reason I'm proposing this change is that it will avoid making a copy of the 'dims' array from oneDNN's tensor to our tensor:
```cpp
struct graph_compiler_tensor {
    size_t id;
    uint8_t ndims;
    size_t dims[12];
    void *data;
};

struct oneDNN_tensor {
    size_t id;
    uint8_t ndims;
    size_t dims[12];
    void *data;
};

void partition_execute(std::vector<oneDNN_tensor>& args) {
    auto onednn_t = args[0];

    // I can't simply assign dims to our_t.dims without copying
    // graph_compiler_tensor our_t {
    //     onednn_t.id,
    //     onednn_t.ndims,
    //     onednn_t.dims, // error: cannot initialize an array element of type 'size_t' (aka 'unsigned long') with an lvalue of type 'size_t[12]' (aka 'unsigned long[12]')
    //     onednn_t.data
    // };
    // I have to make a copy:
    graph_compiler_tensor x{onednn_t.id, onednn_t.ndims, {}, onednn_t.data}; // Initialize dims as empty here
    // Now copy the elements from onednn_t.dims to x.dims
    std::copy(onednn_t.dims, onednn_t.dims + 12, x.dims);
}

int main() {
    // user's code
    oneDNN_tensor t1{0, 3, {1, 2, 3}, nullptr};

    std::vector<oneDNN_tensor> args {t1};

    partition_execute(args);
}
```
If we would use a pointer for dims in our tensors then we'll be able perform the conversion much easier:
```cpp
struct graph_compiler_tensor {
    size_t id;
    uint8_t ndims;
    size_t *dims;
    void *data;
};

struct oneDNN_tensor {
    size_t id;
    uint8_t ndims;
    size_t dims[12];
    void *data;
};

void partition_execute(std::vector<oneDNN_tensor>& args) {
    auto onednn_t = args[0];

    // I can do like this:    
    graph_compiler_tensor our_t {
        onednn_t.id,
        onednn_t.ndims,
        onednn_t.dims,
        onednn_t.data
    };
}

int main() {
    // user's code
    oneDNN_tensor t1{0, 3, {1, 2, 3}, nullptr};

    std::vector<oneDNN_tensor> args {t1};

    partition_execute(args);
}
```
In case of using pointers, we need to make sure though that oneDNN tensors won't die till the execution is complete. If the execution is not async then we can guarantee that, since the tensors won't die until the end of `partition::execute()` (they're passed as arguments), but this can be tricky in the case of async execution. 

@kurapov-peter how graph compiler is expected to perform execution, async or not?

Reply:
Yes, we want it to be able to be async for the GPU case (maybe not that important for CPU). We should give more thought to the graph compiler object btw. It should allow for concurrent compilation and execution.